### ISSUE TITLE
Bulk Transfer Support in IItemHandler

### ISSUE THREAD

**Issue by shartte** [MinecraftForge/MinecraftForge#8175]
## Problem

I'll use the term "large item quantities" for any item stack larger than the stack's max stack size.
Examples for mods dealing in large item quantities are storage drawers, barrels, RS, AE2, bins, and many similar storage systems.

The current `IItemHandler` contract has limited support for such large item quantities by allowing `getStackInSlot` to return a stack with a stack size that exceeds the item's max stack size (See https://github.com/MinecraftForge/MinecraftForge/blob/1.17.x/src/main/java/net/minecraftforge/items/IItemHandler.java#L38).

### Extracting Large Amounts
`extractItem` specifically states that the returned item stack must not exceed the item's maximum stack size.

This has two consequences:

1) it's not possible to simulate the extraction of large item quantities, since splitting the simulation into chunks of the maximum stack size is pointless (every operation will have the same outcome since the state of the underlying inventory does not change inbetween)

2) to actually extract large item quantities, a mod has to loop and extract chunks of max stack size from the same slot

### Inserting Large Amounts
`insertItem` does not specifically state if the inserted stack can exceed the item's maximum stack size or not, so a mod can presumably decide to deal with it or not.

## Proposal

To provide better compatibility between mods that consume or provide large item quantities, I'd propose the following additional methods on IItemHandler. These methods also allow mods to provide optimized versions of what currently are O(n) (where n = slots) operations.

```java
/**
 * Tries to extract as many items that can be stacked with the given template item as requested from this
 * item handler. Will extract regardless of slot and does not respect {@link ItemStack#getMaxStackSize()}.
 *
 * @param template The template item for the items that should be extracted. Only items matching this templates item,
 *             tags and caps will be extracted. {@link ItemStack#EMPTY} is not a valid template.
 *             The handler must not modify this template.
 * @param maxAmount The maximum amount to extract. Must be 0 or greater.
 * @param simulate Whether the operation should actually be performed or only simulated.
 * @return The stack of items that could be extracted. This stack does not adhere to {@link ItemStack#getMaxStackSize()}.
 */
default ItemStack bulkExtractItem(ItemStack template, int maxAmount, boolean simulate) {
    Preconditions.checkArgument(!template.isEmpty(), "what may not be empty");
    Preconditions.checkArgument(maxAmount >= 0, "maxAmount is negative");

    int extracted = 0;
    for (int i = 0; i < getSlots(); ++i) {
        if (ItemHandlerHelper.canItemStacksStack(getStackInSlot(i), template)) {
            extracted += extractItem(i, maxAmount, simulate).getCount();
            if (extracted >= maxAmount) {
                break;
            }
        }
    }

    if (extracted > 0) {
        var result = template.copy();
        result.setCount(extracted);
        return result;
    } else {
        return ItemStack.EMPTY;
    }
}

/**
 * Tries to insert as many of the given item as possible into this item handler, spreading it across
 * slots as needed.
 *
 * @param what The stack that should be inserted. {@link ItemStack#EMPTY} is not valid.
 *             Ownership of this stack is transferred to the item handler.
 *             This stack does not need to adhere to {@link ItemStack#getMaxStackSize()}.
 * @param simulate Whether the operation should actually be performed or only simulated.
 * @return The stack of items that could not be inserted.
 */
default ItemStack bulkInsertItem(ItemStack what, boolean simulate) {
    Preconditions.checkArgument(!what.isEmpty(), "what may not be empty");

    return ItemHandlerHelper.insertItem(this, what, simulate);
}
```

An alternative to implementing this as default methods would be adding a completely different capability for it, but it seems more suited to IItemHandler in general.

I will provide a PR for this in case this proposal not outright rejected.

**Comment by Alpvax** [comment]
I would rather turn this into a more transactional system, where you could extract/insert multiple different items in addition to just multiple stacks.

For example, it could be used to extract 4 gold ingots and 3 gold blocks from a compacting storage drawer. This current approach would not be able to handle that scenario (other than special casing and simulating extracting (3 * 9 + 4) = 31 ingots).
Admittedly it would then become a bigger project to create (and may not be applicable to many inventories), but it would be much more powerful.

Having that as a seperate (ITransactionalItemHandler?) capability could be an alternative approach (with simple wrappers to convert IItemHandlers if desired), but as long as this is an RFC it might be worth considering.

**Comment by shartte** [comment]
@Alpvax This RFC is trying to solve a specific problem without breaking all mods that implement item transfers. I think you should open a new RFC to propose something entirely new.

**Comment by Alpvax** [comment]
My point was that it is an alternative approach to this, just a transaction with only one item.
It would just need a beginTransaction method, returning an IItemTransaction, with a default implementation just using the existing methods.

It's just throwing the idea out for others to discuss, to see if people are more interested in a powerful transaction approach, or just the simple and reliable approach you have proposed.

**Comment by KnightMiner** [comment]
I agree with the sentiment that stack sizes larger than the items max make more sense as multiple operations, so I think its good to bring up. While its a harder system to design, it feels more like an API than a hack using too large stack sizes. Basically, while it may be fine to request more than the max stack size if you logic can handle it, that also opens the door to making the API accept larger stack sizes in `insert` which is now a breaking change, most of my item handlers assume the passed stack is a valid size. i'd rather not have to re-validate the size of the passed stack to ensure it does not exceed it's own stack size. Yes, you could argue your change only applies to `exractItem`, but then you just made the range of possible return values on  `extract` not equal to the domain of valid values for `insertItem`

So overall I feel a proper API for bulk transfer (transactions if you want) is a better solution to the problem you are having, as it leads to a more intuitive API for insertion.

**Comment by Ommina** [comment]
@KnightMiner    But the proposal already uses new method names with default implementations.  Does that not already cover concerns of this being a breaking change?

**Comment by KnightMiner** [comment]
No, because bulk insert and extract both default to calling the existing methods. Bulk insert does nothing to ensure the inserted stack size is not greater than the max, which item handlers are not expecting (they are told the inserted item will never be greater than a valid stack size). I have seen item handler implementations that assume they can directly store a copy of the passed stack if the slot was empty before, size checks are often just used for existing stacks with merging. (you can argue this is a flaw in current implementations, but that implementation is valid according to the current docs)

Bulk extract is possibly calling a simulated insert in a loop, which as stated by the original issue will fail. I suppose it is doing so on different slots each time though, so assuming each slot's insertion is independent (which is assumed by item handler helper already) it probably will not be a problem.

You could arguably get around this by having both bulk methods just force the values to valid stack sizes, those are the best default implementations. If that were the case, this would not seem like as much of a hack.

**Comment by Technici4n** [comment]
> most of my item handlers assume the passed stack is a valid size. i'd rather not have to re-validate the size of the passed stack to ensure it does not exceed it's own stack size.

AE2 and RS have both been calling `insertItem` without clamping to max stack size for years, so unfortunately this is something you should support.

AE2: https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/94ffc4a149678d924fea26ac49f1fbf7c61fd5cb/src/main/java/appeng/me/storage/ItemHandlerAdapter.java#L75
RS: https://github.com/refinedmods/refinedstorage/blob/4cf7062d81dc6a69f647218ab1f7cac9c9ce10aa/src/main/java/com/refinedmods/refinedstorage/apiimpl/storage/externalstorage/ItemExternalStorage.java#L90

Additionally, you can see that in this example we have to iterate through every slot. This PR would allow inventories to optimize this operation. Notably, I'm thinking about ender rift.

**Comment by Technici4n** [comment]
Transactions are a completely different beast, I'm not sure why they're being brought up, except maybe as another possible improvement to IItemHandler? I'm also not sure how you would implement transactions on top of a simulation-based system. In fact, I am sure you can't.

I did implement a transaction-based item and fluid API for Fabric, which makes me pretty qualified to talk about transactions. And I'll be happy to expand on them if you want. In fact, I would suggest you read Fabric's Transfer API before anything else if this is the path you want to investigate. But I'll be happy to clarify some general things regarding transactions.

An important thing to note is that in Fabric's implementation transactions are global to properly allow simulating a compound operation across multiple linked handlers, for example two ender chests with the same frequency. Transactions local to each inventory don't make sense.

An even more important thing to note is that transactions (ones that are implemented in a useful way, such as fabric's) are a completely different paradigm than simulations, and require a serious rethinking of existing inventory interactions. **Forge already has an established API, and I don't think it makes sense to throw it away.** This PR suggests a performance and correctness improvement for the existing API to better support cases where two bulk inventories talk to each other, in a completely opt-in way.

**Comment by KnightMiner** [comment]
I don't see a transaction being impossible or requiring global states, you just need to pass in an object to a simulated request to store the transaction state. As long as you keep passing that item around with all additional requests, the transaction state can be queried and you can make decisions in light of that. It would be complex, especially if placing an item required applying non-stack based changes, leading me to be not sure its a good route. But it would be possible. It was brought up here  because there is a small distinction between bulk item transfer and transactions, both are adding multiple stacks to an inventory at one time. But yes, that belongs in a separate RFC, the usecase of this proposal could be resolved without transactions as is, so transactions are only needed for that problem if this is rejected.

----

> AE2 and RS have both been calling insertItem without clamping to max stack size for years, so unfortunately this is something you should support.

No one has reported issues regarding that and I have had my current implementations for years, so I have as much claim to that being wrong as they have to it being right. Technically more as what I am saying is consistent with the spec and that is not. Not saying I won't support it, just that using the argument "other mods are doing it" does not make it valid.

My main point is even if mods are already doing it, the API says its invalid, so some people assume its invalid and don't check for it. This means it is a breaking change to change the spec as it would break existing mods, the fact existing mods are abusing the spec does not stop this from breaking them.

The proposed solution above could be implemented without changing the spec by splitting the stack before inserting instead of passing a too large stack which prevents this addition from being a breaking change (likewise, AE2 and RS could do the same operation to ensure insertion safety, though I understand this would lessen compatibility with handlers that store more than 1 "stack" in a single slot).

**Comment by Technici4n** [comment]
The problem is that the following happens:
```java
IItemHandler a = // some handler
IItemHandler b = a; // we don't know it's a because we got it from somewhere else, but it is a (or related to it).
// transactions are now impossible/broken on any combined inventory.
IItemHandler combined = new CombinedInvWrapper(a, b);
```
And this is just a simple example. You can make it worse by chaining 1-to-n p2p tunnels and combining that with ender chests. Global transactions avoid this problem because the transaction is shared for all the inventories involved in the transfer operation. Whatever. A separate RFC can be opened for further discussion.

===============================================

> Technically more as what I am saying is consistent with the spec and that is not.

The spec doesn't mention this particularly: https://github.com/MinecraftForge/MinecraftForge/blob/9e476551be4649c35881c103748f3bdc3ca3b353/src/main/java/net/minecraftforge/items/IItemHandler.java#L72. At the end of the day, if you do the check it's just one extra check on your side, but if we split the stack we lose the ability to interact with drawers correctly. So, this is a no-brainer on our side.

While this proposition for bulk operations would indeed solve the issue for our purposes, it can still be quite useful for mods that need per-slot simulation on insert (e.g. thermal dynamics) to ignore the max stack size. I'd go as far as adding to the contract of IItemHandler that stack sizes greater than the item's may be passed. The reference implementation `ItemStackHandler` does indeed perform a check on the stack's max size: https://github.com/MinecraftForge/MinecraftForge/blob/9e476551be4649c35881c103748f3bdc3ca3b353/src/main/java/net/minecraftforge/items/ItemStackHandler.java#L169.

**Comment by Alpvax** [comment]
> Transactions are a completely different beast, I'm not sure why they're being brought up, except maybe as another possible improvement to IItemHandler?

I brought them up because they would be an alternative solution, so if we decide that we want a transaction API, this idea would become obsolete. Therefore they are worth keeping in mind when deciding whether this RFC will progress.

RE: transaction implementation:
You seem to know far more than me on the topic, but I was envisioning a beginTransaction method which would return an object (linked to the item handler) which you could request multiple stacks from, with a "commit" method or something similar to actually alter the items in the inventory.
I haven't thought about it much, so I don't know if that would actually work.

**Comment by shartte** [comment]
There has been no indication that Forge intends to throw `IItemHandler` and its many years of service out the window.
In light of that, I really don't like derailing this proposal by discussing that. Valubable feedback like KnightMiner's might get lost in the noise that way.

**Comment by Alpvax** [comment]
I wasn't suggesting it as a replacement, more as an alternative enhancement. But you're right, this isn't the place to discuss the specifics of it.

**Comment by sciwhiz12** [comment]
@shartte Since Forge 1.18 has just freshly released, it may be a good time to revisit this RFC and create a PR to make the proposed changes, so mods porting to 1.18 can properly make use of them (and to make any necessary breaking changes that this RFC needs to do to existing methods/contracts).
